fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/sorting/FileSorter.java

```diff
@@ -0,0 +1,625 @@
+            }
+        })).thenApply(vignore -> loadResult);
+    }
+
+    @SuppressWarnings("PMD.CompareObjectsWithEquals")
+    private void saveToNextFile(int maxNumFiles) {
+        final long startTime = System.nanoTime();
+        final boolean compress = adapter.isCompressed();
+        final java.security.Key encryptionKey = adapter.getEncryptionKey();
+        Cipher cipher = null;
+        if (!mapSorter.getMap().isEmpty()) {
+            File file;
+            try {
+                file = adapter.generateFilename();
+                try (FileOutputStream fileStream = new FileOutputStream(file)) {
+                    final FileChannel fileChannel = fileStream.getChannel();
+                    final CodedOutputStream headerStream = CodedOutputStream.newInstance(fileStream);
+                    // To stay the same size, field existence must not change.
+                    final RecordSortingProto.SortFileHeader.Builder fileHeader = RecordSortingProto.SortFileHeader.newBuilder()
+                            .setVersion(SORT_FILE_VERSION)
+                            .setNumberOfRecords(0)
+                            .setNumberOfSections(0);
+                    headerStream.writeMessageNoTag(fileHeader.build());
+                    final RecordSortingProto.SortSectionHeader.Builder sectionHeader = RecordSortingProto.SortSectionHeader.newBuilder()
+                            .setNumberOfRecords(0)
+                            .setNumberOfBytes(0);
+                    if (encryptionKey != null) {
+                        cipher = CipherPool.borrowCipher();
+                        initCipherEncrypt(cipher, encryptionKey, adapter.getSecureRandom(), sectionHeader);
+                    }
+                    headerStream.writeMessageNoTag(sectionHeader.build());
+                    final long headerEnd = headerStream.getTotalBytesWritten();
+                    final OutputStream outputStream;
+                    final CodedOutputStream entryStream;
+                    if (compress || cipher != null) {
+                        headerStream.flush();
+                        outputStream = wrapOutputStream(fileStream, cipher, compress);
+                        entryStream = CodedOutputStream.newInstance(outputStream);
+                    } else {
+                        outputStream = fileStream;
+                        entryStream = headerStream;
+                    }
+                    if (timer != null) {
+                        timer.recordSinceNanoTime(SortEvents.Events.FILE_SORT_OPEN_FILE, startTime);
+                    }
+                    int numberOfRecords = 0;
+                    for (Map.Entry<K, V> keyAndValue : mapSorter.getMap().entrySet()) {
+                        final long recordStartTime = System.nanoTime();
+                        entryStream.writeByteArrayNoTag(adapter.serializeKey(keyAndValue.getKey()));
```

Comment:
I don't think the adapter is licensed to encode the key any other way because the sorter code itself needs to understand how it collates, unlike the value, which is opaque.